src/meta/src/hummock/compactor_manager.rs

```diff
@@ -82,164 +92,199 @@ impl Compactor {
         task_expiry_seconds: u64,
     ) -> MetaResult<Self> {
-        let manager = Self::new(task_expiry_seconds);
-        // Initialize the existing task assignments from metastore
-        CompactTaskAssignment::list(env.meta_store())
-            .await?
-            .into_iter()
-            .for_each(|assignment| {
-                manager.initiate_task_heartbeat(
-                    assignment.context_id,
-                    assignment.compact_task.unwrap(),
-                );
-            });
+        // Retrieve the existing task assignments from metastore.
+        let task_assignment = CompactTaskAssignment::list(env.meta_store()).await?;
+        // Initialize the number of tasks assigned to each comapctor.
+        let mut compactor_assigned_task_num = HashMap::new();
+        task_assignment.iter().for_each(|assignment| {
+            compactor_assigned_task_num
+                .entry(assignment.context_id)
+                .and_modify(|n| *n += 1)
+                .or_insert(1);
+        });
+        let manager = Self {
+            policy: RwLock::new(Box::new(ScoredPolicy::new_with_task_assignment(
+                &task_assignment,
+            ))),
+            compactor_assigned_task_num: Mutex::new(compactor_assigned_task_num),
+            task_expiry_seconds,
+            task_heartbeats: Default::default(),
+        };
+        // Initialize heartbeat for existing tasks.
+        task_assignment.into_iter().for_each(|assignment| {
+            manager
+                .initiate_task_heartbeat(assignment.context_id, assignment.compact_task.unwrap());
+        });
         Ok(manager)
     }
 
-    pub fn new(task_expiry_seconds: u64) -> Self {
+    /// Only used for unit test.
+    pub fn new_for_test() -> Self {
         Self {
-            inner: parking_lot::RwLock::new(CompactorManagerInner::new()),
-            task_expiry_seconds,
+            policy: RwLock::new(Box::new(RoundRobinPolicy::new())),
+            compactor_assigned_task_num: Mutex::new(HashMap::new()),
+            task_expiry_seconds: 1,
             task_heartbeats: Default::default(),
         }
     }
 
-    pub async fn next_idle_compactor<S: MetaStore>(
+    /// Only used for unit test.
+    pub fn new_with_policy(policy: Box<dyn CompactionSchedulePolicy>) -> Self {
+        Self {
+            policy: RwLock::new(policy),
+            compactor_assigned_task_num: Mutex::new(HashMap::new()),
+            task_expiry_seconds: 1,
+            task_heartbeats: Default::default(),
+        }
+    }
+
+    /// Gets next idle compactor to assign task.
+    // TODO: If a compactor is returned and `compact_task` is `Some`, then it is considered to be
+    // already assigned to it. This may cause inconsistency between `CompactorManager` and
+    // `HummockManager, if `HummockManager::assign_compaction_task` and `next_idle_compactor` are
+    // not called together. So we might need to put `HummockManager::assign_compaction_task` in this
```

Comment:
Then we don't need to maintain `context_id -> task num` locally anymore, because its purpose was to avoid acquiring the compaction lock in `HummockManager`.